### ISSUE TITLE
[Bug]: mysqlreader : Access denied for user 'root'@'172.17.0.1'

### ISSUE THREAD

**Issue by lschampion** [wgzhao/Addax#561]
### Contact Details（联系人）

911lisa@163.com

### What happened?

mysql reader 读取不到数据库配置。报错

### Version

4.0.8 (Default)

### OS Type

Linux (Default)

### Java JDK Version

Oracle JDK 1.8.0

### Relevant log output

```shell
my json file:
{
        "content":{
                "reader":{
                        "parameter":{
                                "password":"*****",
                                "column":[
                                        "*"
                                ],
                                "connection":[
                                        {
                                                "driver":"com.mysql.jdbc.Driver",
                                                "jdbcUrl":[
                                                        "jdbc:mysql://192.168.100.110:3306/addax_test"
                                                ],
                                                "table":[
                                                        "books"
                                                ]
                                        }
                                ],
                                "where":"1=1",
                                "username":"root"
                        },
                        "name":"mysqlreader"
                },
                "writer":{
                        "parameter":{
                                "password":"*****",
                                "session":[
                                        "set session sql_mode='ANSI'"
                                ],
                                "column":[
                                        "*"
                                ],
                                "connection":[
                                        {
                                                "driver":"com.mysql.jdbc.Driver",
                                                "jdbcUrl":"jdbc:mysql://192.168.100.110:3306/addax_test",
                                                "table":[
                                                        "books_dist"
                                                ]
                                        }
                                ],
                                "writeMode":"insert",
                                "username":"root",
                                "preSql":[
                                        "delete from @table"
                                ]
                        },
                        "name":"mysqlwriter"
                }
        },
        "setting":{
                "speed":{
                        "bytes":-1,
                        "channel":3
                }
        }
}
ERROR:
com.wgzhao.addax.common.exception.AddaxException: Code:[DBUtilErrorCode-10], Description:[连接数据库失败. 请检查您的账号、密码、数据库名称、IP、Port或者向 DBA 寻求帮助(注意网络环境).].  -  具体错误信息为：java.sql.SQLException: Cannot create PoolableConnectionFactory (Access denied for user 'root'@'172.17.0.1' (using password: NO))
```

**Comment by wgzhao** [comment]
请给出完整的报错信息

**Comment by lschampion** [comment]
(base) [root@localhost addax-4.0.7-bin]# bin/addax.sh -d mysql2mysql.json
ready to exec: java -server -Xms64m -Xmx2g -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/root/addax-4.0.7-bin -Dfile.encoding=UTF-8 -Djava.security.egd=file:///dev/urandom -Daddax.home=/root/addax-4.0.7-bin                         -Dlogback.configurationFile=/root/addax-4.0.7-bin/conf/logback.xml  -classpath .:/etc/hbase/conf:/root/addax-4.0.7-bin/lib/*     -Dloglevel=info -Daddax.log=/root/addax-4.0.7-bin/log -Dlog.file.name=addax_mysql2mysql_20220506_084530_46041.log -Xdebug -Xrunjdwp:transport=dt_socket,server=y,address=0.0.0.0:9999 com.wgzhao.addax.core.Engine -job mysql2mysql.json
Listening for transport dt_socket at address: 9999

  ___      _     _
 / _ \    | |   | |
/ /_\ \ __| | __| | __ ___  __
|  _  |/ _` |/ _` |/ _` \ \/ /
| | | | (_| | (_| | (_| |>  <
\_| |_/\__,_|\__,_|\__,_/_/\_\

:: Addax version ::    (v4.0.7)

2022-05-06 08:46:12.935 [        main] INFO  VMInfo               - VMInfo# operatingSystem class => sun.management.OperatingSystemImpl
2022-05-06 08:46:12.948 [        main] INFO  Engine               -
{
        "content":{
                "reader":{
                        "parameter":{
                                "password":"*****",
                                "column":[
                                        "*"
                                ],
                                "connection":[
                                        {
                                                "driver":"com.mysql.jdbc.Driver",
                                                "jdbcUrl":[
                                                        "jdbc:mysql://192.168.100.110:3306/addax_test"
                                                ],
                                                "table":[
                                                        "books"
                                                ]
                                        }
                                ],
                                "where":"1=1",
                                "username":"root"
                        },
                        "name":"mysqlreader"
                },
                "writer":{
                        "parameter":{
                                "password":"*****",
                                "session":[
                                        "set session sql_mode='ANSI'"
                                ],
                                "column":[
                                        "*"
                                ],
                                "connection":[
                                        {
                                                "driver":"com.mysql.jdbc.Driver",
                                                "jdbcUrl":"jdbc:mysql://192.168.100.110:3306/addax_test",
                                                "table":[
                                                        "books_dist"
                                                ]
                                        }
                                ],
                                "writeMode":"insert",
                                "username":"root",
                                "preSql":[
                                        "delete from @table"
                                ]
                        },
                        "name":"mysqlwriter"
                }
        },
        "setting":{
                "speed":{
                        "bytes":-1,
                        "channel":3
                }
        }
}

2022-05-06 08:46:12.978 [        main] INFO  PerfTrace            - PerfTrace traceId=job_-1, isEnable=false, priority=0
2022-05-06 08:46:12.978 [        main] INFO  JobContainer         - Addax jobContainer starts job.
2022-05-06 08:46:12.981 [        main] INFO  JobContainer         - Set jobId = 0
2022-05-06 08:46:13.002 [       job-0] WARN  OriginalConfPretreatmentUtil - use specified driver class: com.mysql.jdbc.Driver
Loading class `com.mysql.jdbc.Driver'. This is deprecated. The new driver class is `com.mysql.cj.jdbc.Driver'. The driver is automatically registered via the SPI and manual loading of the driver class is generally unnecessary.
2022-05-06 08:46:13.729 [       job-0] INFO  OriginalConfPretreatmentUtil - Available jdbcUrl:jdbc:mysql://192.168.100.110:3306/addax_test?yearIsDateType=false&zeroDateTimeBehavior=convertToNull&tinyInt1isBit=false&rewriteBatchedStatements=true.
2022-05-06 08:46:13.730 [       job-0] WARN  OriginalConfPretreatmentUtil - 您的配置文件中的列配置存在一定的风险. 因为您未配置读取数据库表的列，当您的表字段个数、类型有变动时，可能影响任务正确性甚至会运行出错。请检查您的配置并作出修改.
2022-05-06 08:46:13.738 [       job-0] WARN  OriginalConfPretreatmentUtil - use specified driver class: com.mysql.jdbc.Driver
Loading class `com.mysql.jdbc.Driver'. This is deprecated. The new driver class is `com.mysql.cj.jdbc.Driver'. The driver is automatically registered via the SPI and manual loading of the driver class is generally unnecessary.
2022-05-06 08:46:13.955 [       job-0] ERROR RetryUtil            - Exception when calling callable, 异常Msg:Code:[DBUtilErrorCode-10], Description:[连接数据库失败. 请检查您的账号、密码、数据库名称、IP、Port或者向 DBA 寻求帮助(注意网络环境).].  -  具体错误信息为：java.sql.SQLException: Cannot create PoolableConnectionFactory (Access denied for user 'root'@'172.17.0.1' (using password: NO))
com.wgzhao.addax.common.exception.AddaxException: Code:[DBUtilErrorCode-10], Description:[连接数据库失败. 请检查您的账号、密码、数据库名称、IP、Port或者向 DBA 寻求帮助(注意网络环境).].  -  具体错误信息为：java.sql.SQLException: Cannot create PoolableConnectionFactory (Access denied for user 'root'@'172.17.0.1' (using password: NO))
        at com.wgzhao.addax.common.exception.AddaxException.asAddaxException(AddaxException.java:50)
        at com.wgzhao.addax.rdbms.util.RdbmsException.asConnException(RdbmsException.java:40)
        at com.wgzhao.addax.rdbms.util.DBUtil.connect(DBUtil.java:292)
        at com.wgzhao.addax.rdbms.util.DBUtil.lambda$getConnection$2(DBUtil.java:227)
        at com.wgzhao.addax.common.util.RetryUtil$Retry.call(RetryUtil.java:202)
        at com.wgzhao.addax.common.util.RetryUtil$Retry.doRetry(RetryUtil.java:147)
        at com.wgzhao.addax.common.util.RetryUtil.executeWithRetry(RetryUtil.java:59)
        at com.wgzhao.addax.rdbms.util.DBUtil.getConnection(DBUtil.java:227)
        at com.wgzhao.addax.rdbms.util.DBUtil.getConnection(DBUtil.java:220)
        at com.wgzhao.addax.rdbms.util.JdbcConnectionFactory.getConnection(JdbcConnectionFactory.java:52)
        at com.wgzhao.addax.rdbms.writer.util.OriginalConfPretreatmentUtil.dealColumnConf(OriginalConfPretreatmentUtil.java:138)
        at com.wgzhao.addax.rdbms.writer.util.OriginalConfPretreatmentUtil.dealColumnConf(OriginalConfPretreatmentUtil.java:174)
        at com.wgzhao.addax.rdbms.writer.util.OriginalConfPretreatmentUtil.doPretreatment(OriginalConfPretreatmentUtil.java:63)
        at com.wgzhao.addax.rdbms.writer.CommonRdbmsWriter$Job.init(CommonRdbmsWriter.java:68)
        at com.wgzhao.addax.plugin.writer.mysqlwriter.MysqlWriter$Job.init(MysqlWriter.java:62)
        at com.wgzhao.addax.core.job.JobContainer.initJobWriter(JobContainer.java:646)
        at com.wgzhao.addax.core.job.JobContainer.init(JobContainer.java:310)
        at com.wgzhao.addax.core.job.JobContainer.start(JobContainer.java:130)
        at com.wgzhao.addax.core.Engine.start(Engine.java:75)
        at com.wgzhao.addax.core.Engine.entry(Engine.java:132)
        at com.wgzhao.addax.core.Engine.main(Engine.java:163)
2022-05-06 08:46:14.957 [       job-0] ERROR RetryUtil            - Exception when calling callable, 即将尝试执行第1次重试.本次重试计划等待[1000]ms,实际等待[1001]ms, 异常Msg:[Code:[DBUtilErrorCode-10], Description:[连接数据库失败. 请检查您的账号、密码、数据库名称、IP、Port或者向 DBA 寻求帮助(注意网络环境).].  -  具体错误信息为：java.sql.SQLException: Cannot create PoolableConnectionFactory (Access denied for user 'root'@'172.17.0.1' (using password: NO))]
2022-05-06 08:46:16.974 [       job-0] ERROR RetryUtil            - Exception when calling callable, 即将尝试执行第2次重试.本次重试计划等待[2000]ms,实际等待[2001]ms, 异常Msg:[Code:[DBUtilErrorCode-10], Description:[连接数据库失败. 请检查您的账号、密码、数据库名称、IP、Port或者向 DBA 寻求帮助(注意网络环境).].  -  具体错误信息为：java.sql.SQLException: Cannot create PoolableConnectionFactory (Access denied for user 'root'@'172.17.0.1' (using password: NO))]
2022-05-06 08:46:17.003 [       job-0] INFO  StandAloneJobContainerCommunicator - Total 0 records, 0 bytes | Speed 0B/s, 0 records/s | Error 0 records, 0 bytes |  All Task WaitWriterTime 0.000s |  All Task WaitReaderTime 0.000s | Percentage 0.00%
com.wgzhao.addax.common.exception.AddaxException: Code:[DBUtilErrorCode-10], Description:[连接数据库失败. 请检查您的账号、密码、数据库名称、IP、Port或者向 DBA 寻求帮助(注意网络环境).].  -  具体错误信息为：java.sql.SQLException: Cannot create PoolableConnectionFactory (Access denied for user 'root'@'172.17.0.1' (using password: NO))
        at com.wgzhao.addax.common.exception.AddaxException.asAddaxException(AddaxException.java:50)
        at com.wgzhao.addax.rdbms.util.RdbmsException.asConnException(RdbmsException.java:40)
        at com.wgzhao.addax.rdbms.util.DBUtil.connect(DBUtil.java:292)
        at com.wgzhao.addax.rdbms.util.DBUtil.lambda$getConnection$2(DBUtil.java:227)
        at com.wgzhao.addax.common.util.RetryUtil$Retry.call(RetryUtil.java:202)
        at com.wgzhao.addax.common.util.RetryUtil$Retry.doRetry(RetryUtil.java:147)
        at com.wgzhao.addax.common.util.RetryUtil.executeWithRetry(RetryUtil.java:59)
        at com.wgzhao.addax.rdbms.util.DBUtil.getConnection(DBUtil.java:227)
        at com.wgzhao.addax.rdbms.util.DBUtil.getConnection(DBUtil.java:220)
        at com.wgzhao.addax.rdbms.util.JdbcConnectionFactory.getConnection(JdbcConnectionFactory.java:52)
        at com.wgzhao.addax.rdbms.writer.util.OriginalConfPretreatmentUtil.dealColumnConf(OriginalConfPretreatmentUtil.java:138)
        at com.wgzhao.addax.rdbms.writer.util.OriginalConfPretreatmentUtil.dealColumnConf(OriginalConfPretreatmentUtil.java:174)
        at com.wgzhao.addax.rdbms.writer.util.OriginalConfPretreatmentUtil.doPretreatment(OriginalConfPretreatmentUtil.java:63)
        at com.wgzhao.addax.rdbms.writer.CommonRdbmsWriter$Job.init(CommonRdbmsWriter.java:68)
        at com.wgzhao.addax.plugin.writer.mysqlwriter.MysqlWriter$Job.init(MysqlWriter.java:62)
        at com.wgzhao.addax.core.job.JobContainer.initJobWriter(JobContainer.java:646)
        at com.wgzhao.addax.core.job.JobContainer.init(JobContainer.java:310)
        at com.wgzhao.addax.core.job.JobContainer.start(JobContainer.java:130)
        at com.wgzhao.addax.core.Engine.start(Engine.java:75)
        at com.wgzhao.addax.core.Engine.entry(Engine.java:132)
        at com.wgzhao.addax.core.Engine.main(Engine.java:163)

**Comment by lschampion** [comment]
是我的json写错了。


正确Json如下：
{
  "job": {
    "setting": {
      "speed": {
        "channel": 3,
        "bytes": -1
      }
    },
    "content": {
      "reader": {
        "name": "mysqlreader",
        "parameter": {
          "username": "root",
          "password": "123456",
          "column": [
            "*"
          ],
          "connection": [
            {
              "table": [
                "books"
              ],
              "jdbcUrl": [
                "jdbc:mysql://192.168.100.110:3306/addax_test"
              ],
              "driver": "com.mysql.jdbc.Driver"
            }
          ],
          "where": "1=1"
        }
      },
      "writer": {
        "name": "mysqlwriter",
        "parameter": {
          "writeMode": "insert",
          "username": "root",
          "password": "",
          "column": [
            "*"
          ],
          "session": [
            "set session sql_mode='ANSI'"
          ],
          "preSql": [
            "delete from @table"
          ],
          "connection": [
            {
              "jdbcUrl": "jdbc:mysql://192.168.100.110:3306/addax_test",
              "table": [
                "books_dist"
              ],
              "driver": "com.mysql.jdbc.Driver"
            }
          ]
        }
      }
    }
  }
}
